source/iterator/iterator.go

```diff
@@ -0,0 +1,408 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package iterator
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"hash/fnv"
+	"strings"
+
+	"github.com/conduitio-labs/conduit-connector-oracle/coltypes"
+	"github.com/conduitio-labs/conduit-connector-oracle/repository"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"go.uber.org/multierr"
+)
+
+// Iterator represents an implementation of an iterator for Oracle.
+type Iterator struct {
+	repo     *repository.Oracle
+	snapshot *Snapshot
+	cdc      *CDC
+
+	// table represents a table name
+	table string
+	// keyColumn represents a name of column what iterator use for setting key in record
+	keyColumn string
+	// orderingColumn represents a name of column what iterator use for sorting data
+	orderingColumn string
+	// columns represents a list of table's columns for record payload.
+	// if empty - will get all columns
+	columns []string
+	// batchSize represents a size of batch
+	batchSize int
+
+	// columnTypes represents a columns' description from table
+	columnTypes   map[string]coltypes.ColumnDescription
+	hashedTable   uint32
+	trackingTable string
+	snapshotTable string
+}
+
+// Params represents an incoming iterator params for the New function.
+type Params struct {
+	Position       *Position
+	URL            string
+	Table          string
+	KeyColumn      string
+	OrderingColumn string
+	Columns        []string
+	BatchSize      int
+}
+
+// New creates a new instance of the iterator.
+func New(ctx context.Context, params Params) (*Iterator, error) {
+	var err error
+
+	iterator := &Iterator{
+		table:          params.Table,
+		keyColumn:      params.KeyColumn,
+		orderingColumn: params.OrderingColumn,
+		columns:        params.Columns,
+		batchSize:      params.BatchSize,
+	}
+
+	// hash the table name to use it as a postfix in the tracking table and snapshot,
+	// because the maximum length of names (tables, triggers, etc.) is 30 characters
+	h := fnv.New32a()
+	h.Write([]byte(iterator.table))
+	iterator.hashedTable = h.Sum32()
+
+	iterator.snapshotTable = fmt.Sprintf("CONDUIT_SNAPSHOT_%d", iterator.hashedTable)
+	iterator.trackingTable = fmt.Sprintf("CONDUIT_TRACKING_%d", iterator.hashedTable)
+
+	iterator.repo, err = repository.New(params.URL)
+	if err != nil {
+		return nil, fmt.Errorf("new repository: %w", err)
+	}
+
+	// get column types of table for converting
+	iterator.columnTypes, err = coltypes.GetColumnTypes(ctx, iterator.repo, params.Table)
```

Comment:
IIUC, columnTypes, snapshotTable and trackingTable are all used by the snapshot only? If so, it's probably better to have them in the snapshot struct.

Reply:
I've deleted `columnTypes` field from the `Iterator` struct because it can be confusing (these fields are different in Snapshot and CDC).

But I do not agree with deleting `snapshotTable` and `trackingTable` from the `Iterator` struct. I use both of these fields inside the `Iterator` methods.

Reply:
Yup, I noticed they are used in the `Iterator`. IIUC, the `snapshotTable` for example is used only when initializing and closing the `Snapshot` struct, so IMHO all the snapshot-related functionality should in the `Snapshot` struct. Currently we have some of the Snapshot init logic in `Iterator` (init table), and then some of it in the Snapshot struct (loadRows in NewSnapshot).

Reply:
I did this in a separate [PR](https://github.com/conduitio-labs/conduit-connector-oracle/pull/39).